libipl/p10/ipl0.C

```diff
@@ -377,19 +385,84 @@ static int ipl_alignment_check(void)
 static int ipl_set_ref_clock(void)
 {
 	struct pdbg_target *proc;
+    struct pdbg_target *clock_target;
+    enum pdbg_target_status status;
 	int rc = 0;
+    uint8_t clock_count = 0;
 
 	if (ipl_type() == IPL_TYPE_MPIPL)
 		return -1;
 
 	ipl_log(IPL_INFO, "Istep: set_ref_clock: started\n");
 
+	ipl_log(IPL_INFO, "Istep: soft reset clock, and verify clock status register\n");
+	pdbg_for_each_class_target("oscrefclk", clock_target) {
+        uint8_t data;
+
+        status = pdbg_target_probe(clock_target);
+        assert(status == PDBG_TARGET_ENABLED);
+
+        data = OSC_RESET_CMD;
+        rc = i2c_write(clock_target, 0, OSC_CTL_OFFSET, OSC_RESET_CMD_LEN, &data);
+        if(rc){
+            ipl_log(IPL_ERROR, "soft reset command is failed for clock %d, with rc = %d\n",
+                                pdbg_target_index(clock_target), rc);
+
+		    ipl_error_callback(IPL_ERR_CLK_FAILED);
+
+            return 1;
+        }
+        else{
+            ipl_log(IPL_INFO, "soft reset command is successfull for clock %d\n",
+                                pdbg_target_index(clock_target));
+        }
+
+        sleep(1);
+        rc = i2c_read(clock_target, 0, OSC_STAT_OFFSET, OSC_STAT_REG_LEN, &data);
+        if(rc){
+            ipl_log(IPL_ERROR, "status register read is failed for clock %d, with rc = %d\n",
+                                pdbg_target_index(clock_target), rc);
+
+		    ipl_error_callback(IPL_ERR_CLK_FAILED);
+            return 1;
+        }
+        else{
+            ipl_log(IPL_INFO, "status register value for clock %d is 0x%2X\n",
+                                pdbg_target_index(clock_target), data);
+            if(data & OSC_STAT_ERR_MASK){
+                ipl_log(IPL_ERROR, "Calibration is failed for clock %d, status=0x%2X",
+                                pdbg_target_index(clock_target), data);
+                ipl_error_callback(IPL_ERR_CLK_FAILED);
+                return 1;
+            }
+        }
+        clock_count++;
+    }
+
+    if(clock_count != 1 && clock_count != 2){
+        ipl_log(IPL_ERROR, "Invalid number (%d) of clock target found\n",
+                                clock_count);
+        // This is a software bug. So will not call error callback
+        return 1;
+    }
+
 	pdbg_for_each_class_target("proc", proc) {
 		fapi2::ReturnCode fapirc;
 
 		if (!ipl_is_functional(proc))
 			continue;
 
+        if(clock_count == 2){
```

Comment:
What is the behavior of the system if the system has one clock i.e rainier?

Reply:
in this case ATTR_CP_REFCLOCK_SELECT should be equal to OSC0, which is the default value.
Do you want me to set it explicitly

Reply:
Oh, it is the default value.
we don't choose OSC1 in the OpenBMC?
Do you think if it will vary per system then leave it, device tree value will take care if we changed the default for the system?